src/mailadm/conn.py

```diff
@@ -166,28 +149,74 @@ def get_tokeninfo_by_addr(self, addr):
     # user management
     #
 
-    def add_user(self, addr, hash_pw, date, ttl, token_name):
-        self.execute("PRAGMA foreign_keys=on;")
+    def add_email_account_tries(self, token_info, addr=None, password=None, tries=1):
+        """Try to add an email account."""
+        for i in range(tries):
+            try:
+                return self.add_email_account(token_info, addr=addr, password=password)
+            except (MailcowError, DBError):
+                if i + 1 >= tries:
+                    raise
+
+    def add_email_account(self, token_info, addr=None, password=None):
+        """Add an email account to the mailcow server & mailadm
+
+        :param token_info: the token which authorizes the new user creation
+        :param addr: email address for the new account; randomly generated if omitted
+        :param password: password for the new account; randomly generated if omitted
+        :return: a UserInfo object with the database information about the new user, plus password
+        """
+        token_info.check_exhausted()
+        if password is None:
+            password = mailadm.util.gen_password()
+        if addr is None:
+            rand_part = mailadm.util.get_human_readable_id()
+            username = "{}{}".format(token_info.prefix, rand_part)
+            addr = "{}@{}".format(username, self.config.mail_domain)
+        else:
+            if not addr.endswith(self.config.mail_domain):
+                raise ValueError("email {!r} is not on domain {!r}".format(
+                    addr, self.config.mail_domain))
+
+        # first check that mailcow doesn't have a user with that name already:
+        if self.get_mailcow_connection().get_user(addr):
+            raise MailcowError("account %s does already exist" % (addr,))
+
+        self.add_user_db(addr=addr, date=int(time.time()),
+                         ttl=token_info.get_expiry_seconds(), token_name=token_info.name)
+
+        self.log("added addr {!r} with token {!r}".format(addr, token_info.name))
+
+        user_info = self.get_user_by_addr(addr)
+        user_info.password = password
 
-        token = self.get_tokeninfo_by_name(token_name)
-        if token and token.usecount >= token.maxuse:
-            raise TokenExhausted("token {} is exhausted".format(token_name))
+        # seems that everything is fine so far, so let's invoke mailcow:
+        self.get_mailcow_connection().add_user_mailcow(addr, password)
```

Comment:
Consistency between mailadm & mailcow is clear here:

- we first ask mailcow if the user exists already. if yes, we fail.
- we then implicitly ask the db whether the user exists already. if yes, creating it will fail.
- then we create the user in the mailadm.db
- and finally we create it in mailcow; if it fails for other reasons, we roll back the mailadm.db change to ensure consistency again.

Looks safe in my opinion?

Reply:
that's good enough -- there is a possibility that `add_email_account_tries()` will execute `add_email_account` multiple times which could create some zombie users in the db but it's an unlikely edge case and wouldn't cause much damage.

Reply:
oh, how could that happen?